### PR TITLE
Parallel StatefulSet pod create & delete

### LINKED ISSUE #117071: StatefulSet with `podManagementPolicy: Parallel` should create pods in parallel
### What would you like to be added?

When `podManagementPolicy: Parallel` is set, StatefulSetController should issue create requests for replicas in batches (similar to how other controllers work, e.g. ReplicaSetController) when creating pods for the first time and scaling up.

### Why is this needed?

Currently even with `Parallel` podManagementPolicy StatefulSet the replicas are still _created_ sequentially, even though they can start to _run_ in any order.

This severely limits pod throughput when working with large StatefulSets (~thousands of replicas). For example, 10 StatefulSets with 5k pods each achieve a total of ~6 pods/s, while 1k StatefulSets with 50 pods each allows for creating ~120 pods/s. 

`Parallel` policy suggests the pods can be created sequentially (https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetSpec): 

>The alternative policy is Parallel which will create pods in parallel to match the desired scale without waiting, and on scale down will delete all pods at once.

Relevant code (`Parallel` policy controls the value of `monotonic`): https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L424)

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Improves StatefulSet controller performance at creating or removing pods when `Parallel` mode is enabled.

#### Which issue(s) this PR fixes: 
Fixes #117071

#### Does this PR introduce a user-facing change?
```release-note
Faster StatefulSet creation when `Parallel` mode is enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
No change to API or defined behaviour - performance improvement only.

/cc @mborsz

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:465]: If we break here, we should leave the function right after "for _, fn := range fns" as before this PR.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:544]: Ditto: If we break, we must finish function.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:512]: Or maybe we can run this here as well?

This will increase readability, as me (as a code reader), I won't need to remember that there are some actions to execute.

Also from performance perspective, it shouldn't matter much as we don't expect pods to be created and terminated at the same time.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:579]: This code structure suggest we may want to have some helper function like:
func RunInBatches(fns, initialBatchSize).

I wouldn't be much surprised if such function already exists somewhere.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:441]: Isn't this break sufficient to get rid of the "shouldExit" return value?

IIUC we return "true, X" only in two places:
* here
* delete statefulset pod part

In both cases you break right after that so there are no functions in fns added after fn with potential shouldExit = true value.

"if monotonic == true", you have to exist function after executing fns, but before you go to the next loop (e.g. pod deletion) so you can focus on a single for loop.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:399]: Nit: s/fns/actions/ maybe? :)

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:530]: Fixed - added action in this case as well. I think it makes sense for monotonic to preserve the order of logging as well.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:615]: Fixed - added action in this case as well. I think it makes sense for monotonic to preserve the order of logging as well.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:650]: Refactored. The function is based on the one from ReplicaSet controller, but it has to be slightly different because we use different actions here rather than run the same function a given number of times.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:503]: It's not after the fix to above comments about exiting outer function in monotonic mode.

I also think it's easier to follow the code when this value is returned explicitly.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go]: Done

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:583]: For me it's easier to reason about this logic if we gather all actions first and then execute them. I agree it probably won't impact performance much when scaling.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:662]: You probably need to leave function here if you breaked, e.g. on line 629.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:580]: I think the status here may not be up to date -- e.g. if you created some pod it may require updating.

The caller is reading the status even in case of error: https://github.com/kubernetes/kubernetes/blob/cf03aac61973bcfba62721cfeaf713004c7f49c0/pkg/controller/statefulset/stateful_set_control.go#L113-L113

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [reply pkg/controller/statefulset/stateful_set_control.go:615]: I don't think it's fixed -- if we break here we may still be doing some work after like 664 after the change and before PR we haven't reached code there.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:582]: You may want to clear actions, otherwise if we have some actions that don't return shouldexit=true we may execute them multiple times (e.g. createMissingPersistentVolumeClaims).

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:623]: Ooops, I thought this comment referred to a different part of the code. Fixed in this loop as well

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:585]: Done

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:583]: Done

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:677]: Added action returning shouldExit=true in that case

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:393]: Nit: We may consider using named parameters and writing here
```
defer updateStatus(&status, set.Spec.MinReadySeconds, currentRevision, updateRevision, pods)
```

this way, any place we leave the function the update will be always updated.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (mborsz) [inline pkg/controller/statefulset/stateful_set_control.go:569]: Isn't it problematic in parallel mode that here we read replicas[i] and potentially in parallel we create/update that pod?

Or We call UpdateSetatefulPod when the pod is already changed in Delete/CreateStatefulPod.

Also: Does the if in like 566 makes sense if the object can change as part of the action?

Can we maybe instead of adding 10 actions per single pod add one huge action per pod?

Like wrap entire for body in a single action?
High level it should be OK, because we will be doing work for multiple pods in parallel.

It will be also much easier to review.

Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go]: Refactored as suggested. Makes it look like a much smaller change, thanks!

File: pkg/controller/statefulset/stateful_set_control.go
Comment (aleksandra-malinowska) [reply pkg/controller/statefulset/stateful_set_control.go:393]: I thought about this before, but this would also modify status after calling updateStatefulSetAfterInvariantEstablished(), and we don't want to do that.

BTW I think this would work even without named parameters: https://go.dev/play/p/VIqxr6zW6u2

**Comment by aleksandra-malinowska** [comment]
/cc @soltysh 

I think this should be ready for owner review. Lost LGTM from @mborsz after rebase, but hopefully that's a temporary state :)

**Comment by mborsz** [comment]
/lgtm

I believe there are three steps in this change:
* changing status calculation so that we don't track this all the time, rather we recompute at the end
* migrating loop bodies to separate functions
* running the separate functions either sequentially (as in the past) or in parallel (new).

@soltysh if you prefer to see three PRs/commits to make review easier let us know, we can adjust.

File: pkg/controller/statefulset/stateful_set_control.go
Comment (soltysh) [inline pkg/controller/statefulset/stateful_set_control.go:281]: Nice to have: we have this code (either identical named method or identical functionality) spread through several controllers. I counted daemon set, job, replica set, and now this. In a followup, can you extract all of these into single function living in `pkg/controller/controller_utils.go`?

**Review by soltysh — COMMENTED** [review]
This all looks great - thank you! I have a few requests before we can merge this one in:
1. I think it would be nice to split this change into 2 commits (for historical reasons), where first would contain the refactorings related to functions and status calculation, and the second will be the actual changes to make create and delete parallel.
2. Can you add a test (unit or integration, e2e might be unnecessarily long) making sure this new mode is working as expected.

**Comment by aleksandra-malinowska** [comment]
>1. I think it would be nice to split this change into 2 commits (for historical reasons), where first would contain the refactorings related to functions and status calculation, and the second will be the actual changes to make create and delete parallel.

Done

>2. Can you add a test (unit or integration, e2e might be unnecessarily long) making sure this new mode is working as expected.

I've been considering how to best approach this and it looks like we might not be able to validate performance outside of a scalability test suite.

I'm leaning towards a unit test with the following modifications:
- adding an artificial delay to mock API calls,
- checking max number of parallel requests, rather than focusing on timings (otherwise this test would also have to be rather long). 

Would that kind of test be sufficient?

**Comment by mborsz** [comment]
For me such mock test makes more sense than e2e scalability test: it will more isolated, test exactly property we care about, much shorter and cheaper than e2e scalability test.

**Review by soltysh — APPROVED** [review]
/approve

**Comment by nayihz** [comment]
The next pod will wait for `minReadySeconds` after the former one has been created when `podManagementPolicy: OrderedReady`. Does `minReadySeconds` take effect when `podManagementPolicy` is set to `Parallel`?

**Comment by aleksandra-malinowska** [comment]
>The next pod will wait for minReadySeconds after the former one has been created when podManagementPolicy: OrderedReady. Does minReadySeconds take effect when podManagementPolicy is set to Parallel?

`minReadySeconds` is the delay between a pod becoming `Ready` (running and passing readiness probe, if it has one) and being considered available by the StatefulSet controller. This will be reflected in StatefulSet's status regardless of `podManagementPolicy`

In `Parallel` mode, the next pod can be created before the previous one is `Ready` (or even scheduled). This was the case before this change as well.

**Comment by nayihz** [comment]
> In `Parallel` mode, the next pod can be created before the previous one is `Ready` (or even scheduled). This was the case before this change as well.

Yes, that's correct. 
The `minReadySeconds` field does not seem to affect the creation process of a pod in `Parallel` mode.  And the `availableReplicas` seems to be only used for statistical purposes.

**Pull request merged**
